domain/application/state/application.go

```diff
@@ -2193,6 +2194,412 @@ FROM v_revision_updater_application_unit
 	})
 }
 
+// GetApplicationConfig returns the application config attributes for the
+// configuration.
+// If no application is found, an error satisfying
+// [applicationerrors.ApplicationNotFound] is returned.
+func (st *State) GetApplicationConfig(ctx context.Context, appID coreapplication.ID) (map[string]application.ApplicationConfig, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, internalerrors.Capture(err)
+	}
+
+	// We don't currently check for life in the old code, it might though be
+	// worth checking if the application is not dead.
+	ident := applicationID{ID: appID}
+
+	// This isn't ideal, as we could request this in one query, but we need to
+	// perform multiple queries to get the data. First is to get the application
+	// availability, second to just get the application overlay config for the
+	// charm config and the application settings for the trust config.
+	appQuery := `
+SELECT &applicationID.*
+FROM application
+WHERE uuid = $applicationID.uuid;
+`
+	configQuery := `
+SELECT &applicationConfig.*
+FROM v_application_config
+WHERE uuid = $applicationID.uuid;
+`
+	settingsQuery := `
+SELECT &applicationSettings.*
+FROM application_setting
+WHERE application_uuid = $applicationID.uuid;`
+
+	appStmt, err := st.Prepare(appQuery, ident)
+	if err != nil {
+		return nil, internalerrors.Errorf("preparing query for application config: %w", err)
+	}
+	configStmt, err := st.Prepare(configQuery, applicationConfig{}, ident)
+	if err != nil {
+		return nil, internalerrors.Errorf("preparing query for application config: %w", err)
+	}
+	settingsStmt, err := st.Prepare(settingsQuery, applicationSettings{}, ident)
+	if err != nil {
+		return nil, internalerrors.Errorf("preparing query for application config: %w", err)
+	}
+
+	var configs []applicationConfig
+	var settings applicationSettings
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		if err := tx.Query(ctx, appStmt, ident).Get(&ident); errors.Is(err, sqlair.ErrNoRows) {
+			return applicationerrors.ApplicationNotFound
+		} else if err != nil {
+			return internalerrors.Errorf("querying application: %w", err)
+		}
```

Comment:
I'm struggling to understand the intention behind this.

We query the applications to load `LifeID` into `indent`, but then never read the `LifeID` attribute? Why do we do this?

Are we instead just trying to check for existence of an application with this id?

Instead, we should query for `COUNT(*)` and check the count is 1.

This would be much clearer because then explicit that we aren't pulling data out of the db, because we reduce it into a count

Reply:
The contract of the methods, for all methods should be, does the thing you're about to operate on exist? **Not** does the application config return the config values even if they're empty (trapping the sql.ErrNoRows), because the call site compounds this. The call site of these methods does not check if the application exists for the request. So callers, calling the API server will be ill-informed that they can get the configuration of an application that does not exist.

As for the query about life, that was clearly a mistake to override an existing struct to provide additional information.

As for the count query, you could say, that `sql.ErrNoRows` is identical to `COUNT(*) == 1`; does it exist or not? One is more explicit (the count version if you're following along).

Reply:
Agree. The inclusion of `life_id` in the struct confused me, as in isolation, that seemed to be what the code was trying to do, extract that data.

If the life isn't included, it's more obvious you're checking for existence, especially with the error check. however, performing an action to see if it raises an error, and branching based on if it errors or not, I personally find difficult to grok.

So yeah I prefer explicit, `COUNT(*) == 1`

Reply:
I changed it completely now. We check if the application exists in a common method, which also ensures it's not dead.